### ISSUE TITLE
Gradient hook is called twice for shared parameter with activation checkpointing

### ISSUE THREAD

**Issue by szhengac** [pytorch/pytorch#81296]
### 🐛 Describe the bug

One thing I observed is that grad hook has different behaviors when parameter is shared between two layers and activation checkpointing is turned on. When activation checkpointing is not used, grad hook is only called once even this weight is used for multiple times in different layers. However, when activation checkpointing is turned on, grad hook is called multiple times. I don't quite understand why we have such different behaviors.  The following script reproduces the behavior:

```
import torch
from torch.utils import checkpoint

layer1 = torch.nn.Linear(3, 3)
layer2 = layer1 = torch.nn.Linear(3, 3)
layer2.weight =layer1.weight

param = layer1.weight
param_tmp = param.expand_as(param)
grad_acc = param_tmp.grad_fn.next_functions[0][0]

def f(*args):
    print('Gradient hook is called!')

grad_acc.register_hook(f)

x = torch.randn(3, requires_grad=True)
x = layer1(x)
x = layer2(x)
o = x.sum()
o.backward()

print('========================')

def custom():
    def custom_forward(*inputs):
        return layer1(*inputs)
    return custom_forward

x = torch.randn(3, requires_grad=True)
x = checkpoint.checkpoint(custom(), x)
x = layer2(x)
o = x.sum()
o.backward()
```

Output:
```
Gradient hook is called!
========================
Gradient hook is called!
Gradient hook is called!
```

### Versions

PyTorch version: 1.11.0a0+git5f61dde
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Amazon Linux 2 (x86_64)
GCC version: (GCC) 7.3.1 20180712 (Red Hat 7.3.1-15)
Clang version: 7.0.1 (Amazon Linux 2 7.0.1-1.amzn2.0.2)
CMake version: version 3.18.2
Libc version: glibc-2.2.5

Python version: 3.7.10 (default, Jun  3 2021, 00:02:01)  [GCC 7.3.1 20180712 (Red Hat 7.3.1-13)] (64-bit runtime)
Python platform: Linux-4.14.200-155.322.amzn2.x86_64-x86_64-with-glibc2.2.5
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: 
GPU 0: A100-SXM4-40GB
GPU 1: A100-SXM4-40GB
GPU 2: A100-SXM4-40GB
GPU 3: A100-SXM4-40GB
GPU 4: A100-SXM4-40GB
GPU 5: A100-SXM4-40GB
GPU 6: A100-SXM4-40GB
GPU 7: A100-SXM4-40GB

Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.4
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.19.3
[pip3] pytorch-ignite==0.4.2
[pip3] torch==1.11.0a0+git7508557
[pip3] torchvision==0.8.1
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by soulitzer** [comment]
This is expected. Without checkpoint, the weight's post hook is only called once because it needs to wait until all its dependencies are fulfilled. We call the hook after we backprop all the way back to layer1, we don't call the hook as we backprop through layer2 because the dependency from layer1 is not ready yet.

With checkpoint, the weight's hook is called twice because there are actually two backwards that are dealing with two different graphs (that are only somewhat connected). From the perspective of the outer backward, the dependency of first layer on the weight is not built because the custom function does not explicitly take the shared weight as an argument. This means that we actually call the hook before we even reach layer1. The second time we call the hook is during the reentrant backward, which builds a new graph during backward, this graph does have layer1 as a dependency, but doesn't have layer2 as a dependency (we are backwarding from layer1).

**Comment by szhengac** [comment]
@soulitzer thanks for your detailed explanation. Is there any way to identify if this is the last hook called inside the hook function? I am asking that because it helps distributed training for the models that requires shared parameters such as ALBERT and Siamese BERT.

**Comment by soulitzer** [comment]
This doesn't directly answer your question, but you can actually use checkpointing with `use_reentrant=False` and your hook would only be executed once.

**Comment by szhengac** [comment]
@soulitzer can you please explain how `use_reentrant=False` works? And, you said it would be executed only once, is it only executed after all the gradients are accumulated?

**Comment by soulitzer** [comment]
`use_reentrant=False` uses a mechanism called saved_tensor_hooks, which allows you to interpose any logic at the point when tensors are saved and at the point where they are used for backward. This mechanism allows us to achieve the same result of avoiding saving anything except the input for backward, without using custom functions or running a reentrant backward.

> And, you said it would be executed only once, is it only executed after all the gradients are accumulated?

Yes, this avoids the situation I described previously because none of the dependencies would be hidden from the single backward and the hook registered to accumulate grad would only run when all the dependencies have been satisfied.

**Comment by szhengac** [comment]
@soulitzer thanks for the explanation. I just checked the implementation of this mechanism. One thing I don't understand is why this `inner_unpack` function will not be called: https://github.com/pytorch/pytorch/blob/master/torch/utils/checkpoint.py#L363-L364. And I think this approach will discard the input to the forward function as well, which contrasts with the reentrant approach where the inputs are saved?

**Comment by szhengac** [comment]
@soulitzer thanks. I see why `inner_unpack` isn't called now. Though the closure will keep the inputs, the `pack` and `unpack` function are still applied to all the tensors under the context including `*args` and `**kwargs`?

**Comment by soulitzer** [comment]
Yup, the `unpack` function is actually saved on the graph, and since `unpack` references variables in the enclosing scope, those are kept alive as well.

**Comment by szhengac** [comment]
@soulitzer thanks. It is clear now. And I just tried out `reentrant=False`, and it works now.

**Comment by nyngwang** [comment]
> @soulitzer [...] From the perspective of the outer backward, the dependency of first layer on the weight is not built because the custom function does not explicitly take the shared weight as an argument. [...]

Could you elaborate more on this line? I'm trying to learn it but I don't know why it's called the "outer backward"?

**Comment by soulitzer** [comment]
There is an inner and outer backward because in checkpointing's custom Function's backward a nested `.backward` call is made.

**Comment by flyinghu123** [comment]
I tried out reentrant=False and it works now. But in a situation where both True and False can work, setting False results in a 10% performance loss.

**Comment by soulitzer** [comment]
@flyinghu123 

>  I tried out reentrant=False and it works now. But in a situation where both True and False can work, setting False results in a 10% performance loss.

This seems unexpected. If you have more information, feel free to file a separate issue for this.
